ethclient/simulated/backend.go

```diff
@@ -0,0 +1,187 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package simulated
+
+import (
+	"math"
+	"time"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/eth"
+	"github.com/ethereum/go-ethereum/eth/catalyst"
+	"github.com/ethereum/go-ethereum/eth/downloader"
+	"github.com/ethereum/go-ethereum/eth/ethconfig"
+	"github.com/ethereum/go-ethereum/eth/filters"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+// Backend is a simulated blockchain. You can use it to test your contracts or
+// other code that interacts with the Ethereum chain.
+type Backend struct {
+	eth    *eth.Ethereum
+	beacon *catalyst.SimulatedBeacon
+	client simClient
+}
+
+// simClient wraps ethclient. This exists to prevent extracting ethclient.Client
+// from the Client interface returned by Backend.
+type simClient struct {
+	*ethclient.Client
+}
+
+// Client exposes the methods provided by the Ethereum RPC client.
+type Client interface {
+	ethereum.BlockNumberReader
+	ethereum.ChainReader
+	ethereum.ChainStateReader
+	ethereum.ContractCaller
+	ethereum.GasEstimator
+	ethereum.GasPricer
+	ethereum.GasPricer1559
+	ethereum.LogFilterer
+	ethereum.PendingStateReader
+	ethereum.PendingContractCaller
+	ethereum.TransactionReader
+	ethereum.TransactionSender
+	ethereum.ChainIDReader
+}
+
+// New creates a new binding backend using a simulated blockchain
+// for testing purposes.
+// A simulated backend always uses chainID 1337.
+func New(alloc core.GenesisAlloc, gasLimit uint64) *Backend {
+	// Setup the node object
+	nodeConf := node.DefaultConfig
+	nodeConf.DataDir = ""
+	nodeConf.P2P = p2p.Config{NoDiscovery: true}
+	stack, err := node.New(&nodeConf)
+	if err != nil {
+		// This should never happen, if it does, please open an issue
+		panic(err)
+	}
+
+	// Setup ethereum
+	genesis := core.Genesis{
+		Config:   params.AllDevChainProtocolChanges,
+		GasLimit: gasLimit,
+		Alloc:    alloc,
+	}
+	conf := ethconfig.Defaults
+	conf.Genesis = &genesis
+	conf.SyncMode = downloader.FullSync
+	sim, err := NewWithNode(stack, &conf, math.MaxUint64)
+	if err != nil {
+		// This should never happen, if it does, please open an issue
+		panic(err)
+	}
+	return sim
+}
+
+// NewWithNode sets up a simulated backend on an existing node
+// this allows users to do persistent simulations.
+// The provided node must not be started and will be started by NewWithNode
+func NewWithNode(stack *node.Node, conf *eth.Config, blockPeriod uint64) (*Backend, error) {
```

Comment:
How about only taking `dataDir` as parameter here? While custom node can be powerful it can be a way for users to shoot themselves in the foot. E.g. take the node "online" if they forget NoDiscovery.

Reply:
I decided to unexport it for now. If there is demand for other constructors, we should add them in a follow up pr imo.